lib/logstash/outputs/elasticsearch/template_manager.rb

```diff
@@ -42,19 +42,29 @@ def self.install(client, template_endpoint, template_name, template, template_ov
     end
 
     def self.add_ilm_settings_to_template(plugin, template)
+      if plugin.template
+        plugin.deprecation_logger.deprecated("Injecting Index Lifecycle Management configuration into a provided `template` is deprecated, and support will be removed in a future version. Please add the configuration directly to your template.")
```

Comment:
I do agree with most of changes but this message brings one concern to me. We are asking user to `add the configuration directly to their template`. However, we are overwriting their template `index.lifecycle.name` and `index.lifecycle.rollover_alias` ILM settings (`template_manager.rb` L53~L56) with plugin ILM settings, that means user custom template is updated even they add configurations to their template. Both `index.lifecycle.name` and `index.lifecycle.rollover_alias` have default values which may differ from user custom template ILM settings.

I was thinking following two options, they may be applied together but choosing one would simplify:
1) Clear message which should highlight that we will opt out automatic setting but for now, to set same ILM configs int both plugin and template. I would not consider this option as it is manual effort and still under the risk that plugin still automatically applies `index.lifecycle.name` and `index.lifecycle.rollover_alias` to the custom template. 
```suggestion
        plugin.deprecation_logger.deprecated("Injecting Index Lifecycle Management configuration into a provided `template` is deprecated, and support will be removed in a future version. Please add `index.lifecycle.name` and `index.lifecycle.rollover_alias` ILM configurations directly to your template as well as `ilm_policy` and `ilm_rollover_alias` in plugin configs.")
```

2) We automatically set the ILM `index.lifecycle.name` and `index.lifecycle.rollover_alias` _only IFF user custom template doesn't contain these settings_. We also _info_rm user that we are using _custom template ILM configs as it is_ and ignoring plugin level config. With this approach, we achieve backward compatibility. I do believe that the users who touch these settings in the custom template, mostly are knowledgable about ILM policy, rollover, etc..

In my opinion, option-2 looks better approach.
Let me please know if I am missing (or misunderstanding) anything.